pkg/operator/scaling.go

```diff
@@ -0,0 +1,149 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/GoogleCloudPlatform/prometheus-engine/pkg/operator/apis/monitoring/v1"
+	"github.com/go-logr/logr"
+	autoscaling "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	autoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	collectorVPAName = "collector"
+)
+
+type scalingReconciler struct {
+	client client.Client
+	opts   Options
+}
+
+func newScalingReconciler(c client.Client, opts Options) *scalingReconciler {
+	return &scalingReconciler{
+		client: c,
+		opts:   opts,
+	}
+}
+
+func setupScalingController(op *Operator) error {
+	objFilterOperatorConfig := namespacedNamePredicate{
+		namespace: op.opts.PublicNamespace,
+		name:      NameOperatorConfig,
+	}
+
+	err := ctrl.NewControllerManagedBy(op.manager).
+		Named("scaling").
+		WithEventFilter(predicate.ResourceVersionChangedPredicate{}).
+		For(
+			&monitoringv1.OperatorConfig{},
+			builder.WithPredicates(objFilterOperatorConfig),
+		).
+		Owns(&autoscalingv1.VerticalPodAutoscaler{}).
+		Complete(newScalingReconciler(op.manager.GetClient(), op.opts))
+	if err != nil {
+		return fmt.Errorf("scaling controller: %w", err)
+	}
+	return nil
+}
+
+func (r *scalingReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger, _ := logr.FromContext(ctx)
+	logger.WithValues("scaling", req.NamespacedName).Info("reconciling scaling")
+
+	var config monitoringv1.OperatorConfig
+	if err := r.client.Get(ctx, req.NamespacedName, &config); apierrors.IsNotFound(err) {
+		return reconcile.Result{}, deleteVPA(ctx, r.client, r.opts.OperatorNamespace)
+	} else if err != nil {
+		return reconcile.Result{}, fmt.Errorf("get operatorconfig: %w", err)
+	}
+
+	switch {
+	case config.Scaling.VPA.Enabled:
+		// Apply VPA
+		if err := applyVPA(ctx, r.client, r.opts.OperatorNamespace); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	default:
+		return reconcile.Result{}, deleteVPA(ctx, r.client, r.opts.OperatorNamespace)
+	}
+}
+
+func applyVPA(ctx context.Context, c client.Client, namespace string) error {
+	vpa := autoscalingv1.VerticalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      collectorVPAName,
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, c, &vpa, func() error {
```

Comment:
nit: if not needed here, let's just return?

Reply:
The first return value is discarded because we don't particularly care about the OperationResult (whether it was created or updated), as long as it's not an error. However, the parent function does not return an OperationResult. So, we cannot return the result directly.